File: 2023/04.12/서지훈_백준_1149_RGB거리.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());

        int[][] dp = new int[N+1][3];

        for(int i = 1; i <= N; ++i){
            StringTokenizer st = new StringTokenizer(br.readLine());
            int R = Integer.parseInt(st.nextToken());
            int G = Integer.parseInt(st.nextToken());
            int B = Integer.parseInt(st.nextToken());

            dp[i][0] = R + Math.min(dp[i - 1][1], dp[i - 1][2]);
            dp[i][1] = G + Math.min(dp[i - 1][0], dp[i - 1][2]);
            dp[i][2] = B + Math.min(dp[i - 1][0], dp[i - 1][1]);
        }

        bw.write(Integer.toString(Math.min(dp[N][0], Math.min(dp[N][1], dp[N][2]))));

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 2023/04.12/서지훈_백준_1149_RGB거리.kt
import java.io.*
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val N = br.readLine().toInt()

    var prev_house_R_min_cost = 0
    var prev_house_G_min_cost = 0
    var prev_house_B_min_cost = 0

    for(i in 1..N){
        val st = StringTokenizer(br.readLine());
        val R = st.nextToken().toInt()
        val G = st.nextToken().toInt()
        val B = st.nextToken().toInt()

        val R_cost = R + Integer.min(prev_house_G_min_cost, prev_house_B_min_cost)
        val G_cost = G + Integer.min(prev_house_R_min_cost, prev_house_B_min_cost)
        val B_cost = B + Integer.min(prev_house_R_min_cost, prev_house_G_min_cost)

        prev_house_R_min_cost = R_cost
        prev_house_G_min_cost = G_cost
        prev_house_B_min_cost = B_cost
    }

    bw.write(Integer.min(prev_house_R_min_cost, 
        Integer.min(prev_house_G_min_cost, prev_house_B_min_cost)).toString())

    bw.flush()
    bw.close()
    br.close()
}